pil-analyzer/src/structural_checks.rs

```diff
@@ -25,39 +24,61 @@ pub fn check_structs_fields<'a>(
             unreachable!()
         };
 
-        if let Some((
-            _,
-            Some(FunctionValueDefinition::TypeDeclaration(TypeDeclaration::Struct(struct_decl))),
-        )) = definitions.get(name)
-        {
-            let declared_fields: HashSet<_> = struct_decl.fields.iter().map(|f| &f.name).collect();
-            let used_fields: HashSet<_> = fields.iter().map(|f| &f.name).collect();
-
-            errors.extend(
-                fields
-                    .iter()
-                    .filter(|f| !declared_fields.contains(&f.name))
-                    .map(|f| {
-                        sr.with_error(format!("Struct '{name}' has no field named '{}'", f.name))
-                    }),
-            );
-
-            errors.extend(declared_fields.difference(&used_fields).map(|&f| {
-                sr.with_error(format!("Missing field '{f}' in initializer of '{name}'",))
-            }));
-
-            let duplicate_fields: Vec<_> = fields.iter().map(|f| &f.name).duplicates().collect();
-
-            errors.extend(
-                duplicate_fields
-                    .into_iter()
-                    .map(|f| sr.with_error(format!("Field '{f}' specified more than once"))),
-            );
-        }
-
-        visited_structs.insert(name);
+        errors.extend(check_struct_expression(definitions, name, fields, source));
+    }
+
+    errors.extend(check_struct_declarations(definitions));
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+fn check_struct_expression(
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+    name: &String,
+    fields: &[powdr_ast::parsed::NamedExpression<Box<powdr_ast::parsed::Expression<Reference>>>],
+    source: &SourceRef,
+) -> Vec<Error> {
+    let mut errors = Vec::new();
+    if let (
```

Comment:
Shouldn't this panic if it's not a `FunctionValueDefinition::TypeDeclaration`?

Reply:
Since names comes from StructExpressions, None shouldn't be a possible case, hence the direct unwrap.

Reply:
What I meant is that you have an `if let { ... }` here, but there is no else case. Shouldn't it panic in the else case?